Implement MB02ED

saasaa opened the pull request:
Hi,

I have to implement the function MB02ED. Since I am no expert in Fortran, F2Py or numerics I am opening this draft PR to make sure, that I haven't overlooked something obvious or am not following the conventions of this library properly.

I have at the moment:

- [x] Created the F2Py Signature for MB02ED.
- [x] Created the wrapper function `mb02ed` in [slycot/math.py](slycot/math.py)
- [x] Written a test using the example data from [examples](https://github.com/SLICOT/SLICOT-Reference/tree/main/examples) , which passes pytest

I still have to

- [x] Convert the SLICOT Fortran documentation to a numpydocstyle docstring

which I plan to do over the course of this week.

Have I overlooked something?

Thanks

bnavigator commented:
Hi @saasaa,

glad to see someone working on this.

Your seem to have activated some linter in your editor. Please try to avoid the format changes in unrelated functions and import statements at the head of the module files for this PR.

saasaa commented:
I am using black as formatter by default, but I think I have removed all changes.

bnavigator commented on slycot/math.py:
```diff
@@ -23,6 +23,99 @@
+
+    Warns
+    -----
+    SlycotResultWarning
```
These should be errors raising exceptions. And they should be tested.

bnavigator commented on slycot/math.py:
```diff
@@ -23,6 +23,99 @@
+
+    Returns
+    -------
+    X : ndarray
```
B?

bnavigator commented on slycot/math.py:
```diff
@@ -23,6 +23,99 @@
+        Leading nrhs-by-n*k / n*k-by-nrhs part of
+        this array contains the solution matrix X.
+    T: ndarray
+        On exit, if no error is thrown  and  nrhs > 0,  then the leading
```
Strike "on exit".

Also consider handling the dimensions in code.

bnavigator commented on slycot/math.py:
```diff
@@ -23,6 +23,99 @@
+        "info",
+    ]
+
+    T, X, info = _wrapper.mb02ed(typet="C", k=k, n=n, nrhs=nrhs, t=T, b=B)
```
do not silently hardcode `typet` ignoring the wrapper input

saasaa replied:
fixed

bnavigator commented:
Please see #200 and consider using `copy` in the wrapper for `B` and `T`

bnavigator commented:
Don't worry about the current test failures of https://github.com/python-control/Slycot/pull/214/commits/15188ef79ddab48f7100483c56f9d09d0a0417c2 , they look unrelated

saasaa commented:
I have added tests for wrong input parameters and a negative definite matrix, the tests for the wrong parameters pass, but the test for the negative definite matrix gives me an ```TypeError: SlycotError.__init__() takes 3 positional arguments but 4 were given``` and I cannot figure out the cause.

bnavigator commented on slycot/math.py:
```diff
@@ -18,11 +18,113 @@
+        "info",
+    ]
+
+    if k < 0:
+        raise SlycotParameterError(message="k must be >= 0",info=-2)
+    if n< 0:
+        raise SlycotParameterError(message="n must be >= 0", info=-3)
+    if nrhs < 0:
+        raise SlycotParameterError(message="nrhs must be >= 0", info=-4)
```
Not needed. The Fortran routine takes care of it.

https://github.com/python-control/SLICOT-Reference/blob/979f39d7863628407b0f9cae6804efc2833849ab/src/MB02ED.f#L163-L178

saasaa replied:
Can you point me to an example, how to write a test for this case?

bnavigator replied:
The tests are fine, just leave it out of the wrapper itself.

https://github.com/python-control/Slycot/blob/60a5c71190d790b18d62c08b8685e289d75ddf94/slycot/tests/test_tb05ad.py#L155-L173

bnavigator commented:
I fixed the CI errors with setuptools_scm. Please rebase onto current master branch.

bnavigator commented on slycot/math.py:
```diff
@@ -18,11 +18,113 @@
+        k-by-n*k / n*k-by-k part of this array contains the last
+        row / column of the Cholesky factor of inv(T).
+
+    Warns
```
Change to `Raises` so that the docstring parser will generate the correct exception.

https://github.com/python-control/Slycot/blob/60a5c71190d790b18d62c08b8685e289d75ddf94/slycot/exceptions.py#L134

bnavigator commented on slycot/math.py:
```diff
@@ -18,11 +18,106 @@
 
 from . import _wrapper
-from .exceptions import raise_if_slycot_error
+from .exceptions import SlycotParameterError, raise_if_slycot_error
```
"Imported but unused". Please revert this. Technically the error is not raised from any code in this module. (The docstring does not count).

bnavigator commented on slycot/math.py:
```diff
@@ -18,11 +18,106 @@
+        :info = 1:
+            the reduction algorithm failed. The Toeplitz matrix associated
+            with T is not numerically positive definite.
+    SlycotParameterError
+        :info < 0:
+            if INFO = -i, the i-th argument had an illegal value.
```
```
>       with pytest.raises(expected_exception=SlycotParameterError, match="k must be >= 0") as cm:
E       AssertionError: Regex pattern did not match.
E        Regex: 'k must be >= 0'
E        Input: '\nif INFO = -i, the i-th argument had an illegal value.'
```

If you want to have more meaningful error messages (not mandatory);

```suggestion
    SlycotParameterError
        :info = -2:
            k must be >= 0
        :info = -3:
            n must be >= 0
```

and so on

bnavigator replied:
Otherwise just remove the `info < 0` zero case and make sure the `arg_list` is in the correct order, then `raise_if_slycot_error` will generate a reasonable message. You will have to adjust the regex pattern in the tests then.

saasaa commented:
I have added docstrings for the different SlycotParameterError, I have fixed the checks in the function signature, but the tests still fail with ```E       _wrapper.error: (k>=0) failed for 2nd argument k: mb02ed:k=-1```

bnavigator commented on slycot/src/math.pyf:
```diff
@@ -14,9 +14,9 @@ end subroutine mc01td
 subroutine mb02ed(typet,k,n,nrhs,t,ldt,b,ldb,dwork,ldwork,info) ! in MB02ED.f
   character :: typet
-  integer intent(in) :: k
+  integer intent(in),required,check(k>=0)  :: k
```
Too many safeguards for the same thing. Through adding a `check` in the `pyf`, you never get to the Fortran routine generating `INFO=-1`.

bnavigator commented on slycot/math.py:
```diff
@@ -23,6 +23,107 @@
 import numpy as np
 
 
+def mb02ed(typet: str,T: np.ndarray, B: np.ndarray, n: int, k: int,  nrhs: int):
```
```suggestion
def mb02ed(typet: str, T: np.ndarray, B: np.ndarray, n: int, k: int,  nrhs: int):
```

saasaa replied:
May I suggest using black or some other formatter to standardize such formatting questions in the future?

bnavigator replied:
For new code yes. Changing old unrelated code in a feature PR is not so good though.
Also, in the past black did not really work well with reformatting numpy arrays. (https://github.com/python-control/python-control/issues/308)

bnavigator commented on slycot/math.py:
```diff
@@ -23,6 +23,107 @@
+                and the system X*T = B is solved.
+            - 'C': T contains the first block column of an s.p.d. block Toeplitz matrix,
+                and the system T*X = B is solved.
+        Note: the notation x / y means that x corresponds to
+                typet = 'R' and y corresponds to typet = 'C'.
```
remove this or keep its usage below for the dimensions of `T` an `B`

bnavigator commented on slycot/tests/test_mb.py:
```diff
@@ -9,12 +9,184 @@
+    #Test for negative number of right hand sides
+    with pytest.raises(expected_exception=SlycotParameterError, match="nrhs must be >= 0") as cm:
+        mb02ed(T=T, B=B, n=n, k=k, typet=TYPET, nrhs=-1)
+        assert cm.value.info == -4
+
+
+
+def test_mb02ed_matrix_error():
```
```suggestion
        assert cm.value.info == -4


def test_mb02ed_matrix_error():
```

bnavigator commented on slycot/tests/test_mb.py:
```diff
@@ -9,12 +9,184 @@
+    )
+
+    with pytest.raises(SlycotArithmeticError,
+                       match = "\nthe reduction algorithm failed. "
```
```suggestion
                       match = "The reduction algorithm failed. "
```

bnavigator commented on slycot/tests/test_mb.py:
```diff
@@ -9,12 +9,184 @@
+                       "The Toeplitz matrix associated\nwith T "
+                       r"is not numerically positive definite.") as cm:
+        mb02ed(T=T, B=B, n=n, k=k, typet=TYPET, nrhs=nrhs)
+        assert cm.value.info == 1
+
+
+
+
 def test_mb03rd():
```
```suggestion
        assert cm.value.info == 1


def test_mb03rd():
```

bnavigator commented on slycot/math.py:
```diff
@@ -23,6 +23,107 @@
+    ------
+    SlycotArithmeticError
+        :info = 1:
+            the reduction algorithm failed. The Toeplitz matrix associated
```
```suggestion
            The reduction algorithm failed. The Toeplitz matrix associated
```

bnavigator commented on slycot/math.py:
```diff
@@ -23,6 +23,107 @@
+                and the system T*X = B is solved.
+        Note: the notation x / y means that x corresponds to
+                typet = 'R' and y corresponds to typet = 'C'.
+    T : ndarray
+            The leading k-by-n*k or n*k-by-k part of this array must contain the first
+            block row/column of an s.p.d. block Toeplitz matrix.
+    B : ndarray
+            The leading nrhs-by-n*k or n*k-by-nrhs part of this array must contain the
+            right-hand side matrix B.
```
```suggestion
    T : array_like
            The leading k-by-n*k or n*k-by-k part of this array must contain the first
            block row/column of an s.p.d. block Toeplitz matrix.
    B : array_like
            The leading nrhs-by-n*k or n*k-by-nrhs part of this array must contain the
            right-hand side matrix B.
```

bnavigator reviewed (changes requested):
Looks quite good now. A few minor nitpicks:

bnavigator commented:
Thanks a lot for the hard work and the patience with the reviews.

saasaa commented:
Thank you accepting my PR!

bnavigator commented:
Just a heads up @saasaa: You have worked from the `master` branch in your clone. It has now diverged from upstream. You have to reset your `origin` and local `master` branches to `upstream/master` and use a new "feature" branch, when you prepare your next contribution. (https://github.com/python-control/python-control/wiki/How-to-contribute-with-a-pull-request) Looking forward to it.
